Updated declare_permissions (#7249)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/declare_permissions/index.md

> ## Host permissions {: #host-permissions }

Reviewer: The aside warning developers that host permissions trigger a warning was placed on purpose under the host permissions section. Based on developer feedback, many beginner developers are not aware that host permissions trigger a warning, including those declared under content scripts in the manifest. I want to leave the warning in the host permission section in case developers do not continue scrolling down to the Permission Warning section.

Reviewer: What about the use case where developers go directly to the host permissions section from a direct link without starting to read at the very top? Just curious 🤔

Reviewer 2: I apologize for the wording of that. It was harsher than it should have been.

That scenario has exactly the problem I just described. If someone clicks from the top, they have the opportunity to learn about the warnings, but they're probably missing them because they see the link and click through. Readers coming from outside don't even have the opportunity to see that information. 

Another example is [this aside](https://developer.chrome.com/docs/extensions/mv3/declare_permissions/#:~:text=with%20warnings.-,Consider%20using,-optional%20permissions%20wherever) about optional permissions. It links to chrome.permissions, which doesn't appear to even mention what's in the aside. Someone who lands on the chrome.permissions page will never get this information.

I think of asides as for when readers miss things. I'm suggesting that the problem here is that we missed things.

(If we had content for optional host permissions, I'd say make the top a bulleted list and get rid of the definitions in favor of having everything for each manifest key in one place. But this is just a cleanup pass. Maybe later.)

Reviewer 2: I'll open a PR to show you what I have in mind after this one is merged. We can leave the aside in for now.

to them. Consider implementing [optional permissions][api-perms] or a less powerful API to avoid alarming
warnings. For a complete list of best practices, see [Permission warnings guidelines][doc-warning].

Adding or changing match patterns in the `"host_permissions"` and `"content_scripts.matches"` fields will also trigger a [warning](#warnings). To learn more, see [Updating permissions][perm-update].

## Allow access {: #allow_access }
 

> If your extension needs to run on `file://` URLs or operate in incognito mode, users must give the extension access on its details page. 

Reviewer 2: ```suggestion
If your extension needs to run on `file://` URLs or operate in incognito mode, users must give the extension access on its details page. You can find instructions for opening the details page under [Manage your extensions](https://support.google.com/chrome_webstore/answer/2664769?hl=en#:~:text=Manage%20your%20extensions).
```

{% endDetails %}

To detect whether the user has allowed access, you can call [`extension.isAllowedIncognitoAccess()`][incognito-allow] or 
[`extension.isAllowedFileSchemeAccess()`][file-scheme-allow].

## Permissions list {: #permissions }

Here's the list of permissions. See [Permission warnings][doc-warning-table] for a list of permission warnings.

`"accessibilityFeatures.modify"`
: Lets extensions modify accessibility states when using the <a href="/docs/extensions/reference/accessibilityFeatures/"><code>chrome.accessibilityFeatures</code></a> API.

`"accessibilityFeatures.read"`
: Lets extensions read accessibility states when using the <a href="/docs/extensions/reference/accessibilityFeatures/"><code>chrome.accessibilityFeatures</code></a> API.

`"activeTab"`
: Gives temporary access to the active tab through a user gesture. See <a href="/docs/extensions/mv3/manifest/activeTab"><code>activeTab</code></a>
for details.

`"alarms"`
: Gives access to the <a href="/docs/extensions/reference/alarms"><code>chrome.alarms</code></a> API.

> 

Reviewer 2: ```suggestion

`"audio"`
: Gives access to the <a href="/docs/extensions/reference/audio"><code>chrome.audio</code></a> API.
```

`"documentScan"`
: Gives access to the <a href="/docs/extensions/reference/documentScan/">chrome.documentScan</a> API.

`"downloads"`
: Gives access to the <a href="/docs/extensions/reference/downloads/">chrome.downloads</a> API.

`"downloads.open"`
: Allows the use of <a href="/docs/extensions/reference/downloads/#method-open">chrome.downloads.open()</a>.

`"downloads.ui"`
: Allows the use of <a href="/docs/extensions/reference/downloads/#method-setUiOptions">chrome.downloads.setUiOptions()</a>.

`"enterprise.deviceAttributes"`
: Gives access to the <a
          href="/docs/extensions/reference/enterprise_deviceAttributes/">chrome.enterprise.deviceAttributes</a> API.

`"enterprise.hardwarePlatform"`
: Gives access to the <a
          href="/docs/extensions/reference/enterprise_hardwarePlatform/">chrome.enterprise.hardwarePlatform</a> API.

`"enterprise.networkingAttributes"`
: Gives access to the <a        href="/docs/extensions/reference/enterprise_networkingAttributes/">chrome.enterprise.networkingAttributes</a> API.

`"enterprise.platformKeys"`
: Gives access to the <a href="/docs/extensions/reference/enterprise_platformKeys/">chrome.enterprise.platformKeys</a> API.

> 

Reviewer 2: ```suggestion

`"favicon"`
: Grants access to the [Favicon](https://developer.chrome.com/docs/extensions/mv3/favicon/) API.

```

`"offscreen"`
: Gives access to the <a href="/docs/extensions/reference/offscreen/"><code>chrome.offscreen</code></a> API. 

`"pageCapture"`
: Gives access to the <a href="/docs/extensions/reference/pageCapture/">chrome.pageCapture</a> API.

`"platformKeys"`
: Gives access to the <a href="/docs/extensions/reference/platformKeys/">chrome.platformKeys</a> API.

`"power"`
: Gives access to the <a href="/docs/extensions/reference/power/">chrome.power</a> API.

`"printerProvider"`
:Gives access to the <a href="/docs/extensions/reference/printerProvider/">chrome.printerProvider</a> API.

`"printing"`
: Gives access to the <a href="/docs/extensions/reference/printing/">chrome.printing</a> API.

`"printingMetrics"`
: Gives access to the <a href="/docs/extensions/reference/printingMetrics/">chrome.printingMetrics</a> API.

`"privacy"`
: Gives access to the <a href="/docs/extensions/reference/privacy/">chrome.privacy</a> API.

`"processes"`
: Gives access to the <a href="/docs/extensions/reference/processes/">chrome.processes</a> API.

`"proxy"`
: Gives access to the <a href="/docs/extensions/reference/proxy/">chrome.proxy</a> API.

> 

Reviewer 2: ```suggestion

`"runtime"`
: Gives access to the <a href="/docs/extensions/reference/runtime/">runtime</a> API.

```

Reviewer: The runtime API doesn't require any permissions, unless you are using native messaging

https://developer.chrome.com/docs/extensions/reference/runtime/#perms

`"system.memory"`
: Gives access to the <a href="/docs/extensions/reference/system_memory/">chrome.system.memory</a> API.

`"system.storage"`
: Gives access to the <a href="/docs/extensions/reference/system_storage/">chrome.system.storage</a> API.

`"tabCapture"`
: Gives access to the <a href="/docs/extensions/reference/tabCapture/">chrome.tabCapture</a> API.

`"tabGroups"`
: Gives access to the <a href="/docs/extensions/reference/tabGroups/">chrome.tabGroups</a> API.

`"tabs"`
: Gives access to privileged fields of the <a
          href="/docs/extensions/reference/tabs#type-Tab"><code>Tab</code></a> objects used by several APIs,
including <a href="/docs/extensions/reference/tabs">chrome.tabs</a> and <a href="/docs/extensions/reference/windows">chrome.windows</a>. You usually won't need to declare this permission to use those APIs.

`"topSites"`
: Gives access to the <a href="/docs/extensions/reference/topSites/">chrome.topSites</a> API.

`"tts"`
: Gives access to the <a href="/docs/extensions/reference/tts/">chrome.tts</a> API.

`"ttsEngine"`
: Gives access to the <a href="/docs/extensions/reference/ttsEngine/">chrome.ttsEngine</a> API.

> `"unlimitedStorage"`

Reviewer: @oliverdunk I was under the impression that `unlimitedStorage` only worked with `chrome.storage.local` not sync, managed or sessions...

I noticed this also implied in https://developer.chrome.com/docs/extensions/mv3/storage-and-cookies/#storage-persistence

Did this change at some point?

Reviewer 3: Oh interesting, I hadn't given that thought but it looks like you're right - this applies to only `chrome.storage.local` (which makes sense). It also applies to web storage like IndexedDB, where it also exempts that storage from eviction.

We should probably rewrite this with that in mind :)

Reviewer 2: @oliverdunk see the line below.

Reviewer 3: I don't think this says it is limited to chrome.storage.local, right? What we have now ("an unlimited quota for chrome.storage") feels misleading even if we do have clarification.

Reviewer 2: Sorry. Next two lines. It says:

"Unlimited quota for chrome.storage, IndexedDB, Cache Storage, and Origin Private File System.

Reviewer 3: That's still wrong though, right? There is not "unlimited quota for chrome.storage", only `chrome.storage.local`.

Reviewer 2: >It also applies to web storage like IndexedDB, where it also exempts that storage from eviction.

Please list all types of memory that `unlimitedStorage` applies to.

Reviewer 3: IndexedDB, Cache Storage and WebSQL.

Author: so the change here is to replace chrome.storage and Origin Private File System with WebSQL?

Reviewer 3: I would say:

- Add WebSQL to list including Origin Private File System (apologies, I forgot about that one, but it looks like it would apply).
- Replace `chrome.storage` with `chrome.storage.local`

Reviewer 2: Looks good.

Author: ok! What links do I need to add?

Reviewer 2: - As Oliver said, change `chrome.storage` to `chrome.storage.local` and change the link to this: https://pr-7249-static-dot-dcc-staging.uc.r.appspot.com/docs/extensions/reference/storage/#property-local
- Remove the links to IndexedDB and Cache Storage.

@oliverdunk, can we have a sidebar discussion about WebSQL? It's deprecated, but some reason, I have it in the back of my head that this is sticking around for the extensions universe. I'd like to let Alex finish this PR, and it'll take me all of five minutes to fix this later.

Reviewer 3: > Remove the links to IndexedDB and Cache Storage.

I'm not sure I agree with this? It affects both of those APIs.

> @oliverdunk, can we have a sidebar discussion about WebSQL? It's deprecated, but some reason, I have it in the back of my head that this is sticking around for the extensions universe.

For sure - TLDR, it's going away for everyone. There's currently a deprecation trial to get it back if you need it but that won't last forever. I called it out for completeness but could also see an argument not to document it.

Reviewer 2: @oliverdunk I am super confused. Can you please just give us a list of every API covered by `"unlimitedStorage"` unlimited storage?

Reviewer 3: `chrome.storage.local`, IndexedDB, Cache Storage, Origin Private File System and WebSQL.

Reviewer 2: As Oliver and I discussed yesterday, we need to leave WebSQL off the list. It's deprecated. We don't want anyone doing anything new with it.

`"identity"`
: Gives access to the <a href="/docs/extensions/reference/identity/">chrome.identity</a> API.

`"idle"`
: Gives access to the <a href="/docs/extensions/reference/idle/">chrome.idle</a> API.

`"loginState"`
: Gives access to the <a href="/docs/extensions/reference/loginState/">chrome.loginState</a> API.

`"management"`
: Gives access to the <a href="/docs/extensions/reference/management/">chrome.management</a> API.

`"nativeMessaging"`
: Gives access to the <a href="/docs/extensions/mv3/nativeMessaging/">native messaging API</a>.

`"notifications"`
: Gives access to the <a href="/docs/extensions/reference/notifications/">chrome.notifications</a> API.

`"offscreen"`
: Gives access to the <a href="/docs/extensions/reference/offscreen/"><code>chrome.offscreen</code></a> API. 

`"pageCapture"`
: Gives access to the <a href="/docs/extensions/reference/pageCapture/">chrome.pageCapture</a> API.

`"platformKeys"`
: Gives access to the <a href="/docs/extensions/reference/platformKeys/">chrome.platformKeys</a> API.

`"power"`
: Gives access to the <a href="/docs/extensions/reference/power/">chrome.power</a> API.

`"printerProvider"`

> :Gives access to the <a href="/docs/extensions/reference/printerProvider/">chrome.printerProvider</a> API.

Reviewer 2: ```suggestion
: Gives access to the <a href="/docs/extensions/reference/printerProvider/">chrome.printerProvider</a> API.
```

{% endDetails %}

To detect whether the user has allowed access, you can use [`extension.isAllowedIncognitoAccess()`][incognito-allow] or 
[`extension.isAllowedFileSchemeAccess()`][file-scheme-allow].

## Permissions list {: #permissions }

The following table lists the currently available permissions. See Permission warnings][doc-warning-table] for a list of permission warnings.

`"activeTab"`
: Gives temporary access to the active tab through a user gesture. See <a href="/docs/extensions/mv3/manifest/activeTab"><code>activeTab</code></a>
for details.

`"alarms"`
: Gives access to the <a href="/docs/extensions/reference/alarms">chrome.alarms</a> API.

`"background"`

> : Makes Chrome start up early (as soon as the user logs into their computer, before they launch Chrome), and shut down late (even after its last window is closed, until the user explicitly quits Chrome).

Reviewer 3: It means background in the "run in the background" sense :)

> : Makes Chrome start up early (as soon as the user logs into their computer, before they launch Chrome), and shut down late (even after its last window is closed, until the user explicitly quits Chrome).

Reviewer: It can still be used by service workers in MV3 👍

> : Makes Chrome start up early (as soon as the user logs into their computer, before they launch Chrome), and shut down late (even after its last window is closed, until the user explicitly quits Chrome).

Reviewer 2: I had that question because 'background' was overloaded. Might an extension developer have the same question? Is there anything we can do to shore this up?

`"declarativeContent"`
: Gives access to the <a href="/docs/extensions/reference/declarativeContent/">chrome.declarativeContent</a> API.

`"declarativeNetRequest"`
: Gives your extension access to the <a href="/docs/extensions/reference/declarativeNetRequest/">chrome.declarativeNetRequest</a> API. Some operations also require host permissions.

`"declarativeNetRequestWithHostAccess"`
: Gives extension access to the <a href="/docs/extensions/reference/declarativeNetRequest/">chrome.declarativeNetRequest</a> API. Also requires host permissions to the request URL and initiator to act on a request.

`"declarativeNetRequestFeedback"`
: Gives access to events and methods within the <a
          href="/docs/extensions/reference/declarativeNetRequest/">chrome.declarativeNetRequest</a> API, which returns information on declarative rules matched.

<!-- No corresponding reference entry
    `"displaySource"`
    : Gives access to the <a href="/docs/extensions/reference/displaySource/">chrome.displaySource</a> API.

    `"dns"`
    : Gives access to the <a href="/docs/extensions/reference/dns/">chrome.dns</a> API.
-->

> <!--Do these still need to be commented out?-->

Reviewer: `dns` can be uncommented. It has it's own reference page
https://developer.chrome.com/docs/extensions/reference/dns/

`"enterprise.hardwarePlatform"`
: Gives access to the <a
          href="/docs/extensions/reference/enterprise_hardwarePlatform/">chrome.enterprise.hardwarePlatform</a> API.

`"enterprise.networkingAttributes"`
: Gives access to the <a        href="/docs/extensions/reference/enterprise_networkingAttributes/">chrome.enterprise.networkingAttributes</a> API.

`"enterprise.platformKeys"`
: Gives access to the <a href="/docs/extensions/reference/enterprise_platformKeys/">chrome.enterprise.platformKeys</a> API.

`"experimental"`
: Allows the use of <a href="/docs/extensions/reference/#experimental_apis/">chrome.experimental.* APIs</a>.

`"fileBrowserHandler"`
: Gives access to the <a href="/docs/extensions/reference/fileBrowserHandler/">chrome.fileBrowserHandler</a> API.

`"fileSystemProvider"`
: Gives access to the <a href="/docs/extensions/reference/fileSystemProvider/">chrome.fileSystemProvider</a> API.

`"fontSettings"`
: Gives access to the <a href="/docs/extensions/reference/fontSettings/">chrome.fontSettings</a> API.

`"gcm"`
: Gives access to the <a href="/docs/extensions/reference/gcm/">chrome.gcm</a> API.

`"geolocation"`
:Allows the extension to use the <a

>           href="https://dev.w3.org/geo/api/spec-source.html">geolocation API</a> without prompting the user for permission.

Reviewer 2: ```suggestion
          href="https://developer.mozilla.org/docs/Web/API/Geolocation_API">geolocation API</a> without prompting the user for permission.
```

Reviewer 2: If you find any references to w3c or wecg specs, please look for them on [MDN](https://developer.mozilla.org/) and use that link instead. The links you copy will contain a country code ('en-US/'). Please remove that from the URL.

`"history"`
:Gives access to the <a href="/docs/extensions/reference/history/">chrome.history</a> API.

`"identity"`
: Gives access to the <a href="/docs/extensions/reference/identity/">chrome.identity</a> API.

`"idle"`
: Gives access to the <a href="/docs/extensions/reference/idle/">chrome.idle</a> API.

> 

Reviewer 2: @oliverdunk We're missing the permission for `input.ime`. That's going away isn't it?

Reviewer 3: I think it makes sense to keep documented for now. At most it's deprecated, and there are definitely still extensions that rely on it.

`"tabGroups"`
: Gives access to the <a href="/docs/extensions/reference/tabGroups/">chrome.tabGroups</a> API.

`"tabs"`
: Gives access to privileged fields of the <a
          href="/docs/extensions/reference/tabs#type-Tab"><code>Tab</code></a> objects used by several APIs,
including <a href="/docs/extensions/reference/tabs">chrome.tabs</a> and <a href="/docs/extensions/reference/windows">chrome.windows</a>. You usually won't need to declare this permission to use those APIs.

`"topSites"`
: Gives access to the <a href="/docs/extensions/reference/topSites/">chrome.topSites</a> API.

`"tts"`
: Gives access to the <a href="/docs/extensions/reference/tts/">chrome.tts</a> API.

`"ttsEngine"`
: Gives access to the <a href="/docs/extensions/reference/ttsEngine/">chrome.ttsEngine</a> API.

`"unlimitedStorage"`
: Provides unlimited storage for clientside data to Web SQL Database and application caches <!--has http://crbug.com/58985 been fixed?-->. Without this permission, the extension is limited to 10 MB of local storage.
<!--This permission applies only to Web SQL Database and application cache

>           (see issue <a href="http://crbug.com/58985">58985</a>). Also, it doesn't currently work with wildcard

Reviewer 4: It's on track, yes.

>           (see issue <a href="http://crbug.com/58985">58985</a>). Also, it doesn't currently work with wildcard

Author: Done.
